README.md
![1337 Readme Logo](.github/images/1337-readme.svg)

Share your progress, skills, stats with the world via GitHub `README.md`

[![code style: prettier](https://img.shields.io/badge/code%20style-prettier-ff69b4)](https://github.com/prettier/prettier)
[![code style: prettier](https://img.shields.io/badge/gitmoji-%20😜%20😍-FFDD67.svg)](https://gitmoji.dev)
[![license: MIT](https://img.shields.io/badge/license-MIT-green)](LICENSE)
[![Vercel](https://badgen.net/badge/icon/vercel/black?icon=vercel&label)](https://vercel.com)<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg)](#contributors-)

<!-- ALL-CONTRIBUTORS-BADGE:END -->

---

## Preview

[![42 Profile Card](https://1337-readme.vercel.app/api/profile?cursus=42cursus&login=azorkane)](https://github.com/mohouyizme/1337-readme)

[![42 Profile Card](https://1337-readme.vercel.app/api/profile?dark=true&login=toxiic)](https://github.com/mohouyizme/1337-readme)

## Usage

1. Go to [1337-readme.vercel.app](https://1337-readme.vercel.app) and generat code snippet for your profile card.
2. Create a repository that's the same name as your GitHub username. For example, my GitHub username is `mohouyizme` so I created a new repository with the name `mohouyizme`. The letter-casing must match your GitHub username.
3. Create a `README.md` file inside the repository.
4. Paste the code snippet into the `README.md` file.
5. Enjoy 🎉!

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://ouyiz.me"><img src="https://avatars1.githubusercontent.com/u/23297439?v=4" width="100px;" alt=""/><br /><sub><b>Mohamed Ouyizme</b></sub></a><br /><a href="https://github.com/mohouyizme/1337-readme/commits?author=mohouyizme" title="Code">💻</a> <a href="https://github.com/mohouyizme/1337-readme/issues?q=author%3Amohouyizme" title="Bug reports">🐛</a> <a href="#content-mohouyizme" title="Content">🖋</a> <a href="https://github.com/mohouyizme/1337-readme/commits?author=mohouyizme" title="Documentation">📖</a> <a href="#design-mohouyizme" title="Design">🎨</a> <a href="#ideas-mohouyizme" title="Ideas, Planning, & Feedback">🤔</a> <a href="#maintenance-mohouyizme" title="Maintenance">🚧</a> <a href="#platform-mohouyizme" title="Packaging/porting to new platform">📦</a> <a href="https://github.com/mohouyizme/1337-readme/pulls?q=is%3Apr+reviewed-by%3Amohouyizme" title="Reviewed Pull Requests">👀</a></td>
    <td align="center"><a href="https://stackoverflow.com/users/2595450/mehdi-bounya"><img src="https://avatars2.githubusercontent.com/u/5004111?v=4" width="100px;" alt=""/><br /><sub><b>Mehdi Bounya</b></sub></a><br /><a href="https://github.com/mohouyizme/1337-readme/commits?author=mehdibo" title="Code">💻</a> <a href="#ideas-mehdibo" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://reversablecode.github.io"><img src="https://avatars3.githubusercontent.com/u/56657108?v=4" width="100px;" alt=""/><br /><sub><b>Abdessamad El bahri</b></sub></a><br /><a href="https://github.com/mohouyizme/1337-readme/commits?author=ReversableCode" title="Code">💻</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

---

[CHANGELOG](CHANGELOG.md) - [LICENSE](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Skills card.
- Project score/reviews.

## [1.1.0] - 2020-12-23

### Changed

- 1337 Readme now is rewritten in TypeScript 🎉 (Thanks to @ReversableCode).
- You can write case insensitive logins in the website (Thanks to @mehdibo).

## [1.0.5] - 2020-12-19

### Added

- Add animation to profile card.

### Changed

- Enhance spacing on profile card logos.
- Change `CHANGLOG` versions order.

## [1.0.4] - 2020-12-15

### Added

- Add cursus name to profile card.
- Add `Piscine C décloisonnée` to cursuses list.

### Changed

- Add `Pisciner` tag to piscine candidates.

## [1.0.3] - 2020-12-15

### Added

- Add usage section to README.md

### Changed

- Fix profile card layout issues on mobile.

## [1.0.2] - 2020-12-13

### Changed

- Fix background issues on profile card.

## [1.0.1] - 2020-12-13

### Added

- Cache profiles and images for 2 hours.
- Add a CHANGELOG.md to this project 😜.


next-env.d.ts
/// <reference types="next" />
/// <reference types="next/types/global" />

interface FortyTwoCursusData {
  id: number
  name: string
  slug: string
}
interface FortyTwoCursus {
  cursus_id: number
  grade: string | null
  level: number
  cursus: FortyTwoCursusData
}
interface FortyTwoUser {
  email: string
  first_name: string
  last_name: string
  image_url: string
  'staff?': boolean
  cursus_users: FortyTwoCursus[]
}
interface FortyTwo {
  getUser(login: string): Promise<FortyTwoUser>
}
interface NextApiCustomRequest extends NextApiRequest {
  fortyTwo: FortyTwo
}

interface FortyTwoCursusData {
  id: number
  name: string
  slug: string
}
interface FortyTwoCursus {
  cursus_id: number
  grade: string | null
  level: number
  cursus: FortyTwoCursusData
}
interface FortyTwoUser {
  email: string
  first_name: string
  last_name: string
  image_url: string
  'staff?': boolean
  cursus_users: FortyTwoCursus[]
}
interface QueryData {
  login: string
  cursus: string
  email: string
  leet_logo: string
  forty_two_network_logo: string
  dark: string
}
interface CustomRequest extends IncomingMessage {
  fortyTwo: FortyTwo
  query: QueryData
}

interface UserProfileCursus {
  grade: string | null
  name: string
  level: number
}
interface UserProfile {
  login: string
  fullName: string
  email: string | boolean
  image: string
  isPisciner: boolean
  isStaff: boolean
  cursus: UserProfileCursus | boolean
  leetLogo: boolean
  fortyTwoLogo: boolean
  isDark: boolean
}


pages/index.js
import Head from 'next/head'

import Logo from '../components/Logo'
import ProfileCardGenerator from '../components/ProfileCardGenerator'

export default function HomePage() {
  return (
    <>
      <Head>
        <script async defer src="https://buttons.github.io/buttons.js" />
      </Head>
      <div className="max-w-screen-md mx-auto">
        <div className="flex flex-col items-center my-20">
          <Logo />
          <p className="my-5">
            Share your progress, skills, stats with the world via GitHub
            README.md
          </p>
          <a
            className="github-button"
            href="https://github.com/mohouyizme/1337-readme"
            data-color-scheme="no-preference: dark; light: dark; dark: dark;"
            data-show-count="true"
            aria-label="Star mohouyizme/1337-readme on GitHub"
          >
            Star
          </a>
        </div>
        <ProfileCardGenerator />
      </div>
    </>
  )
}


middlewares/42.ts
import axios from 'axios'
import { NextApiResponse } from 'next'
import { NextHandler } from 'next-connect'

const { FORTY_TWO_ENDPOINT, FORTY_TWO_UID, FORTY_TWO_SECRET } = process.env

export default function fortyTwo(
  req: NextApiCustomRequest,
  _res: NextApiResponse,
  next: NextHandler
) {
  req.fortyTwo = { getUser }
  next()
}

async function getToken() {
  const response = await axios.post(`${FORTY_TWO_ENDPOINT}/oauth/token`, {
    grant_type: 'client_credentials',
    client_id: FORTY_TWO_UID,
    client_secret: FORTY_TWO_SECRET,
  })
  return response.data
}

async function getUser(login: string): Promise<FortyTwoUser> {
  const { access_token } = await getToken()
  const response = await axios.get(`${FORTY_TWO_ENDPOINT}/v2/users/${login}`, {
    headers: {
      Authorization: `Bearer ${access_token}`,
    },
  })
  return response.data
}


pages/api/profile.js
import nc from 'next-connect'
import cache from 'memory-cache'
import imageToBase64 from 'image-to-base64'
import { renderToStaticMarkup } from 'react-dom/server'

import fortyTwo from '../../middlewares/42'
import Profile from '../../components/cards/Profile'

const handler = nc()

handler.use(fortyTwo)

handler.get(async (req, res) => {
  const {
    login,
    cursus,
    email: queryEmail,
    leet_logo,
    forty_two_network_logo,
    dark,
  } = req.query

  let userData
  const userCache = cache.get(`login:${login}`)

  if (userCache) userData = userCache
  else {
    userData = await req.fortyTwo.getUser(login)
    cache.put(`login:${login}`, userData, 7200000)
  }

  const {
    email,
    first_name,
    last_name,
    image_url,
    'staff?': isStaff,
    cursus_users: cursuses,
  } = userData

  const getCursus = cursuses.find(({ cursus: { slug } }) => slug === cursus)
  const isPisciner = cursuses.length === 1 && cursuses[0].cursus_id === 6
  let image
  const imageCache = cache.get(`image:${login}`)
  if (imageCache) image = imageCache
  else {
    image = await imageToBase64(image_url)
    cache.put(`image:${login}`, image, 7200000)
  }

  const user = {
    login,
    fullName: `${first_name} ${last_name}`,
    email: queryEmail !== 'hide' && email,
    image,
    isPisciner,
    isStaff,
    cursus: !isStaff &&
      getCursus && {
        grade: getCursus.grade,
        name: getCursus.cursus.name,
        level: getCursus.level,
      },
    leetLogo: leet_logo !== 'hide',
    fortyTwoLogo: forty_two_network_logo !== 'hide',
    isDark: dark === 'true',
  }

  res.setHeader('Content-Type', 'image/svg+xml')
  res.end(renderToStaticMarkup(<Profile user={user} />))
})

export default handler


components/ProfileCardGenerator.js
import queryString from 'query-string'
import { Formik } from 'formik'
import { useState } from 'react'

export default function ProfileCardGenerator() {
  const [cardImage, setCardImage] = useState(
    '/api/profile?login=ozaazaa&cursus=42&email=hide'
  )

  function generateProfileCard({
    login,
    cursus,
    showEmail,
    darkTheme,
    hideLeetLogo,
    hideFortyTwoNetworkLogo,
  }) {
    const query = { login: login.toLowerCase() }
    if (cursus !== 'hide') query.cursus = cursus
    if (!showEmail) query.email = 'hide'
    if (darkTheme) query.dark = 'true'
    if (hideLeetLogo) query.leet_logo = 'hide'
    if (hideFortyTwoNetworkLogo) query.forty_two_network_logo = 'hide'
    const queryStringified = queryString.stringify(query)
    setCardImage(`/api/profile?${queryStringified}`)
  }

  return (
    <div>
      <h3 className="mb-5 text-xl font-bold">Profile Card Generator</h3>
      <div className="flex">
        <div className="mr-5">
          <img src={cardImage} width={450} />
          <p className="mt-5">
            Add the code bellow to your{' '}
            <code className="px-2 py-1 text-sm rounded bg-gray-50">
              README.md
            </code>{' '}
            file:
          </p>
          <div
            className="p-3 mt-2 overflow-x-auto font-mono text-xs rounded-lg select-all bg-gray-50"
            style={{ maxWidth: 450 }}
          >
            [![42 Profile Card](https://1337-readme.vercel.app{cardImage}
            )](https://github.com/mohouyizme/1337-readme)
          </div>
        </div>
        <Formik
          initialValues={{
            login: 'ozaazaa',
            cursus: '42',
            showEmail: false,
            darkTheme: false,
            hideLeetLogo: false,
            hideFortyTwoNetworkLogo: false,
          }}
          onSubmit={(values) => {
            setCardImage(null)
            generateProfileCard(values)
          }}
        >
          {({ values, handleChange, handleBlur, handleSubmit }) => (
            <form onSubmit={handleSubmit}>
              <div className="mb-5">
                <label
                  className="inline-block mb-2 font-bold cursor-pointer"
                  htmlFor="login"
                >
                  login
                </label>
                <input
                  className="w-full px-6 py-3 transition rounded bg-gray-50 focus:outline-none focus:ring focus:bg-white"
                  type="text"
                  name="login"
                  id="login"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.login}
                />
              </div>
              <div className="mb-5">
                <label
                  className="inline-block mb-2 font-bold cursor-pointer"
                  htmlFor="cursus"
                >
                  cursus
                </label>
                <div className="relative">
                  <select
                    className="w-full px-6 py-3 transition rounded appearance-none pr-9 bg-gray-50 focus:outline-none focus:bg-white focus:ring"
                    name="cursus"
                    id="cursus"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.cursus}
                  >
                    <option value="hide">hide</option>
                    <option value="42">42</option>
                    <option value="42cursus">42cursus</option>
                    <option value="piscine-c-decloisonnee">
                      Piscine C décloisonnée
                    </option>
                  </select>
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    fill="currentColor"
                    aria-hidden="true"
                    className="absolute w-6 h-6 pointer-events-none right-2 top-3"
                    viewBox="0 0 20 20"
                  >
                    <path
                      fillRule="evenodd"
                      d="M10 3a1 1 0 01.707.293l3 3a1 1 0 01-1.414 1.414L10 5.414 7.707 7.707a1 1 0 01-1.414-1.414l3-3A1 1 0 0110 3zm-3.707 9.293a1 1 0 011.414 0L10 14.586l2.293-2.293a1 1 0 011.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
                      clipRule="evenodd"
                    ></path>
                  </svg>
                </div>
              </div>
              <div className="mb-2">
                <label className="inline-flex items-center cursor-pointer select-none">
                  <input
                    type="checkbox"
                    className="w-5 h-5 cursor-pointer"
                    name="showEmail"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    checked={values.showEmail}
                  />
                  <span className="ml-3 text-gray-700">Show email</span>
                </label>
              </div>
              <div className="mb-2">
                <label className="inline-flex items-center cursor-pointer select-none">
                  <input
                    type="checkbox"
                    className="w-5 h-5 cursor-pointer"
                    name="darkTheme"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    checked={values.darkTheme}
                  />
                  <span className="ml-3 text-gray-700">Dark theme</span>
                </label>
              </div>
              <div className="mb-2">
                <label className="inline-flex items-center cursor-pointer select-none">
                  <input
                    type="checkbox"
                    className="w-5 h-5 cursor-pointer"
                    name="hideLeetLogo"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    checked={values.hideLeetLogo}
                  />
                  <span className="ml-3 text-gray-700">Hide 1337 logo</span>
                </label>
              </div>
              <div className="mb-2">
                <label className="inline-flex items-center cursor-pointer select-none">
                  <input
                    type="checkbox"
                    className="w-5 h-5 cursor-pointer"
                    name="hideFortyTwoNetworkLogo"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    checked={values.hideFortyTwoNetworkLogo}
                  />
                  <span className="ml-3 text-gray-700">
                    Hide 42 Network logo
                  </span>
                </label>
              </div>
              <button
                type="submit"
                className="w-full p-3 mt-2 text-sm font-bold text-white bg-black rounded focus:outline-none focus:ring"
              >
                Generate
              </button>
            </form>
          )}
        </Formik>
      </div>
    </div>
  )
}
